pygmt/tests/test_clib_put_vector.py

```diff
@@ -0,0 +1,132 @@
+"""
+Test the functions that put vector data into GMT.
+"""
+import numpy as np
+import numpy.testing as npt
+import pytest
+
+from .. import clib
+from ..exceptions import GMTCLibError, GMTInvalidInput
+from ..helpers import GMTTempFile
+
+
+def test_put_vector():
+    "Check that assigning a numpy array to a dataset works"
+    dtypes = "float32 float64 int32 int64 uint32 uint64".split()
+    for dtype in dtypes:
+        with clib.Session() as lib:
+            dataset = lib.create_data(
+                family="GMT_IS_DATASET|GMT_VIA_VECTOR",
+                geometry="GMT_IS_POINT",
+                mode="GMT_CONTAINER_ONLY",
+                dim=[3, 5, 1, 0],  # columns, rows, layers, dtype
+            )
+            x = np.array([1, 2, 3, 4, 5], dtype=dtype)
+            y = np.array([6, 7, 8, 9, 10], dtype=dtype)
+            z = np.array([11, 12, 13, 14, 15], dtype=dtype)
+            lib.put_vector(dataset, column=lib["GMT_X"], vector=x)
+            lib.put_vector(dataset, column=lib["GMT_Y"], vector=y)
+            lib.put_vector(dataset, column=lib["GMT_Z"], vector=z)
+            # Turns out wesn doesn't matter for Datasets
+            wesn = [0] * 6
+            # Save the data to a file to see if it's being accessed correctly
+            with GMTTempFile() as tmp_file:
+                lib.write_data(
+                    "GMT_IS_VECTOR",
+                    "GMT_IS_POINT",
+                    "GMT_WRITE_SET",
+                    wesn,
+                    tmp_file.name,
+                    dataset,
+                )
+                # Load the data and check that it's correct
+                newx, newy, newz = tmp_file.loadtxt(unpack=True, dtype=dtype)
+                npt.assert_allclose(newx, x)
+                npt.assert_allclose(newy, y)
+                npt.assert_allclose(newz, z)
+
+
+def test_put_vector_mixed_dtypes():
+    """Passing a numpy array of mixed dtypes to a dataset.
+    See https://github.com/GenericMappingTools/pygmt/issues/255
+    """
+    dtypes = "float32 float64 int32 int64 uint32 uint64".split()
+    for dtypex in dtypes:
+        for dtypey in dtypes:
+            # skip same dtypes
+            if dtypex == dtypey:
+                continue
```

Comment:
The Python standard library has an [`itertools.combination`](https://docs.python.org/3.8/library/itertools.html#itertools.combinations) function to do exactly this. You'll need to add `import itertools` at the top first, and then change the following lines to this:

```suggestion
    for dtypex, dtypey in itertools.combinations(iterable=dtypes, r=2):
```

Reply:
Great to learn new things everyday!

But it's slightly different from what I want. This is an example:

```python
In [8]: for x, y in itertools.combinations([1, 2, 3], r=2):
   ...:     print(x, y)
   ...:
1 2
1 3
2 3
```
but I want 
```
1 2 
1 3
2 1
2 3
3 1
3 2
```

Reply:
Use [`itertools.permutations`](https://docs.python.org/3.8/library/itertools.html#itertools.permutations) then? I don't think the order matters here, unless there's a module I'm not aware of that's picky about this.

Reply:
The idea of the test is to verify the GMT bug fixed in https://github.com/GenericMappingTools/gmt/issues/3330. As I understand it, the order matters.

`itertools.permutations` is exactly what I need. Will make the change later.